nano/node/bootstrap/bootstrap_lazy.cpp

```diff
@@ -0,0 +1,598 @@
+		else if (lazy_undefined_links.find (next_block.link) == lazy_undefined_links.end ())
+		{
+			lazy_add (next_block.link, node->network_params.bootstrap.lazy_retry_limit); // Head is not confirmed. It can be account or hash or non-existing
+			lazy_undefined_links.insert (next_block.link);
+		}
+		lazy_state_backlog.erase (find_state);
+	}
+}
+
+void nano::bootstrap_attempt_lazy::lazy_backlog_cleanup ()
+{
+	uint64_t read_count (0);
+	auto transaction (node->store.tx_begin_read ());
+	for (auto it (lazy_state_backlog.begin ()), end (lazy_state_backlog.end ()); it != end && !stopped;)
+	{
+		if (node->store.block_exists (transaction, it->first))
+		{
+			auto next_block (it->second);
+			if (node->ledger.balance (transaction, it->first) <= next_block.balance) // balance
+			{
+				lazy_add (next_block.link, next_block.retry_limit); // link
+			}
+			else
+			{
+				lazy_destinations_increment (next_block.link);
+			}
+			it = lazy_state_backlog.erase (it);
+		}
+		else
+		{
+			lazy_add (it->first, it->second.retry_limit);
+			++it;
+		}
+		// We don't want to open read transactions for too long
+		++read_count;
+		if (read_count % batch_read_size == 0)
+		{
+			transaction.refresh ();
+		}
+	}
+}
+
+void nano::bootstrap_attempt_lazy::lazy_destinations_increment (nano::account const & destination_a)
+{
+	// Enabled only if legacy bootstrap is not available. Legacy bootstrap is a more effective way to receive all existing destinations
+	if (node->flags.disable_legacy_bootstrap)
+	{
+		// Update accounts counter for send blocks
+		auto existing (lazy_destinations.get<account_tag> ().find (destination_a));
+		if (existing != lazy_destinations.get<account_tag> ().end ())
+		{
+			lazy_destinations.get<account_tag> ().modify (existing, [](nano::lazy_destinations_item & item_a) {
+				++item_a.count;
+			});
+		}
+		else
+		{
+			lazy_destinations.emplace (nano::lazy_destinations_item{ destination_a, 1 });
+		}
+	}
+}
+
+void nano::bootstrap_attempt_lazy::lazy_destinations_flush ()
+{
+	assert (!mutex.try_lock ());
+	lazy_destinations_flushed = true;
+	size_t count (0);
+	for (auto it (lazy_destinations.get<count_tag> ().begin ()), end (lazy_destinations.get<count_tag> ().end ()); it != end && count < nano::bootstrap_limits::lazy_destinations_request_limit && !stopped;)
+	{
+		lazy_add (it->account, node->network_params.bootstrap.lazy_destinations_retry_limit);
+		it = lazy_destinations.get<count_tag> ().erase (it);
+		++count;
+	}
+}
+
+void nano::bootstrap_attempt_lazy::lazy_blocks_insert (nano::block_hash const & hash_a)
+{
+	assert (!mutex.try_lock ());
+	auto inserted (lazy_blocks.insert (std::hash<::nano::block_hash> () (hash_a)));
+	if (inserted.second)
+	{
+		++lazy_blocks_count;
+	}
+}
+
+void nano::bootstrap_attempt_lazy::lazy_blocks_erase (nano::block_hash const & hash_a)
+{
+	assert (!mutex.try_lock ());
+	auto erased (lazy_blocks.erase (std::hash<::nano::block_hash> () (hash_a)));
+	if (erased)
+	{
+		--lazy_blocks_count;
```

Comment:
Worth having an `assert` than it is greater than 0?

Reply:
It can become 0 so I guess we can check if it's not equal to `std::numeric_limits<size_t>::max ()` (less than 0 for uint)

Reply:
I meant 0 before deletion, but that works too :)